src/solvers/strings/string_format_builtin_function.cpp

```diff
@@ -0,0 +1,489 @@
+    inputs,
+    generator.array_pool,
+    // TODO: get rid of this argument
+    messaget{message_handler},
+    generator.ns);
+  INVARIANT(
+    simplify_expr(result_constraint_pair.first, generator.ns).is_zero(),
+    "add_axioms_for_format should return 0, meaning that formatting was"
+    "successful");
+  result_constraint_pair.second.existential.push_back(
+    equal_exprt{return_code, from_integer(0, get_return_code_type())});
+  return result_constraint_pair.second;
+}
+
+/// Return an new expression representing the length of the representation of
+/// `integer`.
+/// If max_length is less than the length of `integer`, the returned
+/// expression will represent max_length.
+/// \param pos_integer: positive (but not necessarily constant) integer for
+///                     which to compute the length of the decimal
+///                     representation.
+/// \param max_length: max length of the decimal representation. If `max_length`
+///                    is less than the length of `integer`, the returned
+///                    expression will represent `max_length`. Choosing a value
+///                    greater than the actual max possible length is harmless
+///                    but will result in useless constraints.
+/// \param length_type: type to give to the created expression
+/// \param radix: radix of the output representation.
+/// \return: An `exprt` representing the length of `integer`
+static exprt length_of_positive_decimal_int(
+  const exprt &pos_integer,
+  int max_length,
+  const typet &length_type,
+  const unsigned long radix)
+{
+  if(max_length <= 1)
+    return from_integer(1, length_type);
+
+  // If the current value doesn't fit in a smaller size representation, we have
+  // found the number of digits needed to represent that value.
+  const mp_integer max_value_for_smaller_size =
+    pow((mp_integer)radix, max_length - 1);
+  return if_exprt{
+    less_than(
+      pos_integer,
+      from_integer(max_value_for_smaller_size, pos_integer.type())),
+    length_of_positive_decimal_int(
+      pos_integer, max_length - 1, length_type, radix),
+    from_integer(max_length, length_type)};
+}
+
+/// Compute the length of the decimal representation of an integer.
+/// \param integer: (not necessarily constant) integer for which to compute the
+///                 length of the decimal representation.
+/// \param length_type: type to give to the created expression
+/// \param radix: radix of the output representation.
+/// \return: An `exprt` representing the length of `integer`
+exprt length_of_decimal_int(
+  const exprt &integer,
+  const typet &length_type,
+  const unsigned long radix)
+{
+  int max_pos_int_length;
+  if(length_type == signedbv_typet(32))
+  {
+    if(radix == 10)
+      max_pos_int_length = 10;
+    else if(radix == 16)
+      max_pos_int_length = 8;
+    else
+      UNIMPLEMENTED;
+  }
+  else
+  {
+    // We only handle 32-bit signed integer type for now.
```

Comment:
Is there a ticket that describes these limitations?

Reply:
For your first comment, this case is now handled (fixed in the fixup commits). For the second one, I don't think this directly translates to a limitation of CBMC. We would only come across it if we venture in supporting String.format for other languages than Java.

Reply:
There is no ticket for the other limitations. Do you think we need a ticket for each of them? That would be more than 30 tickets I think.

Reply:
It's fine to have a summary ticket or some other form of documentation.